File: Iteration3/Python/Transcript.py
class Transcript(Student):
    def __init__(self, student_id, first_name=None, last_name=None, term=None, gpa=None, cgpa=None,
                 credit_received=None, credit_earned=None, total_credit_received=None, total_credit_earned=None,
                 passed_courses=None, failed_courses=None):
        super().__init__(student_id, first_name, last_name)
        self.gpa = gpa
        self.cgpa = cgpa
        self.credit_received = credit_received
        self.credit_earned = credit_earned
        self.total_credit_received = total_credit_received
        self.total_credit_earned = total_credit_earned
        self.passed_courses = passed_courses if passed_courses else []
        self.failed_courses = failed_courses if failed_courses else []
        self.choosable_courses = []

    def is_prerequisite_courses_passed(self, course):
        temp = 0
        for prerequisite in course.view_prerequisite_courses():
            if self.is_passed_course(prerequisite):
                temp += 1

            for failed_course in self.failed_courses:
                if (failed_course.get_course_letter_grade() == "FD" and
                        prerequisite.get_course_id().get_id() == failed_course.get_course_id().get_id()):
                    temp += 1

        return temp == len(course.view_prerequisite_courses())

    def is_passed_course(self, course):
        for passed_course in self.passed_courses:
            if passed_course.get_course_id().get_id() == course.get_course_id().get_id():
                return True
        return False

    def is_failed_course(self, course):
        for failed_course in self.failed_courses:
            if failed_course.get_course_id().get_id() == course.get_course_id().get_id():
                return True
        return False

    def is_course_taken_before(self, course):
        return self.is_passed_course(course) or self.is_failed_course(course)

    def is_choosable_course(self, course):
        self.set_choosable_courses()
        return course in self.choosable_courses

    def is_chosen_course(self, course_section):
        return course_section in self.get_chosen_courses()

    def get_gpa(self):
        return self.gpa

    def set_gpa(self, gpa):
        self.gpa = gpa

    def get_cgpa(self):
        return self.cgpa

    def set_cgpa(self, cgpa):
        self.cgpa = cgpa

    def get_credit_received(self):
        return self.credit_received

    def set_credit_received(self, credit_received):
        self.credit_received = credit_received

    def get_credit_earned(self):
        return self.credit_earned

    def set_credit_earned(self, credit_earned):
        self.credit_earned = credit_earned

    def get_total_credit_received(self):
        return self.total_credit_received

    def set_total_credit_received(self, total_credit_received):
        self.total_credit_received = total_credit_received

    def get_total_credit_earned(self):
        return self.total_credit_earned

    def set_total_credit_earned(self, total_credit_earned):
        self.total_credit_earned = total_credit_earned

    def view_successful_courses(self):
        print("Successful Courses:")
        for course in self.passed_courses:
            print(f"{course.get_course_name()} - {course.get_course_letter_grade()}")

    def view_failed_courses(self):
        print("Failed Courses:")
        for course in self.failed_courses:
            print(f"{course.get_course_name()} - {course.get_course_letter_grade()}")

    def get_passed_courses(self):
        return self.passed_courses

    def set_passed_courses(self, passed_courses):
        self.passed_courses = passed_courses

    def get_failed_courses(self):
        return self.failed_courses

    def set_failed_courses(self, failed_courses):
        self.failed_courses = failed_courses

    def get_choosable_courses(self):
        return self.choosable_courses

    def set_choosable_courses(self):
        self.choosable_courses = []
        for course in ObjectCreator.courses:
            if (course.get_course_term() <= self.get_term() and not self.is_passed_course(course) and
                    self.is_prerequisite_courses_passed(course)):
                self.choosable_courses.append(course)
            elif (self.is_failed_course(course) and any(
                    failed_course.get_course_letter_grade() == "FD" for failed_course in self.failed_courses)):
                self.choosable_courses.append(course)

File: Iteration3/src3(Java)/Transcript.java
import java.util.ArrayList;

public class Transcript extends Student{
	
	private double gpa;
	private double cgpa;
	private int creditReceived;
	private int creditEarned;
	private int totalCreditReceived;
	private int totalCreditEarned;
    private ArrayList<Course> passedCourses;
    private ArrayList<Course> failedCourses;
    private ArrayList<Course> choosableCourses;
	
	public Transcript(ID StudentID) {
		super(StudentID);
	}
	
	public Transcript(ID StudentID, String firstName, String lastName, int term, double gpa,
			double cgpa, int creditReceived, int creditEarned, int totalCreditReceived, 
			int totalCreditEarned, ArrayList<Course> passedCourses, ArrayList<Course> failedCourses) {
		super(StudentID);
		this.setGpa(gpa);
		this.setCpga(cgpa);
		this.setCreditReceived(creditReceived);
		this.setCreditEarned(creditEarned);
		this.setCreditReceived(totalCreditReceived);
		this.setTotalCreditEarned(totalCreditEarned);
		this.setPassedCourses(passedCourses);
		this.setFailedCourses(failedCourses);
		}

  public boolean isPrerequisiteCoursesPassed(Course course) {	
	int temp = 0;
	for(int i = 0; i < course.viewPrerequisiteCourses().size(); i++) {
		if (isPassedCourse(course.viewPrerequisiteCourses().get(i))){
		temp++;
		}
		for(int j = 0; j < this.failedCourses.size(); j++) {
			// FD is not a passing grade, however with FD student can take the next course
			if(this.failedCourses.get(j).getCourseLetterGrade().equals("FD")
					&& course.viewPrerequisiteCourses().get(i).getCourseID().getId().equals(this.failedCourses.get(j).getCourseID().getId())) {
				temp++;
			}
		}
	}	
    if (temp == course.viewPrerequisiteCourses().size()) {
    	return true;
    }
    else {
    	return false;
    }
}
	
  public boolean isPassedCourse(Course course) {
	  for(int i = 0; i < this.passedCourses.size(); i++) {
	  if (this.passedCourses.get(i).getCourseID().getId().equals(course.getCourseID().getId())) {
		   return true;
	  	}
	  }
	  return false;
}
  
  public boolean isFailedCourse(Course course) {
      for (int i = 0; i < this.getFailedCourses().size(); i++) {
          if (this.getFailedCourses().get(i).getCourseID().getId().equals(course.getCourseID().getId())) {
              return true;
          }
      }
      return false;
  }
	
  public boolean isCourseTakenBefore(Course course) {
  for (int i = 0; i < ObjectCreator.courses.size(); i++) {
      if (isPassedCourse(course) || isFailedCourse(course)) {
          return true;
      }
  }
  return false;
}
	
  public boolean isChoosableCourse(Course course) {
	   this.setChoosableCourses();
    for (int i = 0; i < ObjectCreator.courses.size(); i++) {
        if (this.getChoosableCourses().contains(course)) {
            return true;
        }
    }
    return false;
}

public boolean isChosenCourse(CourseSection cs) {
	   this.setChoosableCourses();
    for (int i = 0; i < this.getChosenCourses().size(); i++) {
        if (this.getChosenCourses().contains(cs)) {
            return true;
        }
    }
    return false;
}
	

	public double getGpa() {
		return gpa;
	}

	public void setGpa(double gpa) {
		this.gpa = gpa;
	}

	public double getCgpa() {
		return cgpa;
	}

	public void setCpga(double cgpa) {
		this.cgpa = cgpa;
	}

	public int getCreditReceived() {
		return creditReceived;
	}

	public void setCreditReceived(int creditReceived) {
		this.creditReceived = creditReceived;
	}

	public int getCreditEarned() {
		return creditEarned;
	}

	public void setCreditEarned(int creditEarned) {
		this.creditEarned = creditEarned;
	}

	public int getTotalCreditReceived() {
		return totalCreditReceived;
	}

	public void setTotalCreditReceived(int totalCreditReceived) {
		this.totalCreditReceived = totalCreditReceived;
	}

	public int getTotalCreditEarned() {
		return totalCreditEarned;
	}

	public void setTotalCreditEarned(int totalCreditEarned) {
		this.totalCreditEarned = totalCreditEarned;
	}
	
	public void viewSuccessfulCourses() {
		System.out.println("Succesful Courses: ");
		for (int i = 0; i < this.getPassedCourses().size(); i++) {
			System.out.println(this.getPassedCourses().get(i).getCourseName() +
					" - " + this.getPassedCourses().get(i).getCourseLetterGrade());
		}
		
	}
	
	public void viewFailedCourses() {
		System.out.println("Failed Courses: ");
		for (int i = 0; i < this.getFailedCourses().size(); i++) {
			System.out.println(this.getFailedCourses().get(i).getCourseName() +
					" - " + this.getFailedCourses().get(i).getCourseLetterGrade());
			
		}
		
	}

	public ArrayList<Course> getPassedCourses() {
		return passedCourses;
	}

	public void setPassedCourses(ArrayList<Course> passedCourses) {
		this.passedCourses = passedCourses;
	}

	public ArrayList<Course> getFailedCourses() {
		return failedCourses;
	}

	public void setFailedCourses(ArrayList<Course> failedCourses) {
		this.failedCourses = failedCourses;
	}
	
	public ArrayList<Course> getChoosableCourses() {
		return this.choosableCourses;
	}
	
	public void setChoosableCourses() {
		choosableCourses = new ArrayList<Course>();
		boolean temp = false;
		for (int i = 0; i < ObjectCreator.courses.size(); i++) {
			// each course in all courses is evaluated. if one's term is lower than student's and one is in the student's passed courses list
			if (((ObjectCreator.courses.get(i).getCourseTerm() <= this.getTerm()) && !(isPassedCourse(ObjectCreator.courses.get(i)))
					&& isPrerequisiteCoursesPassed(ObjectCreator.courses.get(i)))) {
				temp = true;
				}
			for (int j = 0; j < this.getFailedCourses().size(); j++)
			if (isFailedCourse(ObjectCreator.courses.get(i)) && this.getFailedCourses().get(j).getCourseLetterGrade().equals("FD")){
				temp = true;
				
			}
				// each course in all courses is evaluated. if one's prerequisite is not passed
				if (temp == true) {
				this.choosableCourses.add(ObjectCreator.courses.get(i));
					}
				temp = false;
				}
			}
		

}